### ISSUE TITLE
[PATCH] stop Win32 VC miniperl from exporting functions

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12480]
Migrated from [rt.perl.org#115216](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=115216) (status was 'resolved')

Searchable as RT115216$

**Comment by p5pRT** [comment]
### From @bulk88

#### Created by @bulk88



Patch coming soon\. I need to get a RT number first\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.17.5:

Configured by Owner at Mon Oct  8 03:07:26 2012.

Summary of my perl5 (revision 5 version 17 subversion 5) configuration:

  Platform:
    osname=MSWin32, osvers=5.1, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -GF -W3 -MD -Zi -DNDEBUG -O1 -G7 -GL
-DWIN32 -D_CONSOLE -DNO_STRICT  -DPERL_TEXTMODE_SCRIPTS
-DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DUSE_PERLIO
-D_USE_32BIT_TIME_T',
    optimize='-MD -Zi -DNDEBUG -O1 -G7 -GL',
    cppflags='-DWIN32'
    ccversion='13.10.6030', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=8
    ivtype='long', ivsize=4, nvtype='double', nvsize=8,
Off_t='__int64', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf
-ltcg  -libpath:"c:\perl517\lib\CORE"  -machine:x86'
    libpth="C:\Program Files\Microsoft Visual Studio .NET 2003\VC7\lib"
    libs=oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib
netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib  version.lib
odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib
winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib
oleaut32.lib  netapi32.lib uuid.lib ws2_32.lib mpr.lib winmm.lib
version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=true, libperl=perl517.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug
-opt:ref,icf -ltcg  -libpath:"c:\perl517\lib\CORE"  -machine:x86'

Locally applied patches:



@INC for perl 5.17.5:
    C:/perl517/site/lib
    C:/perl517/lib
    .


Environment for perl 5.17.5:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\perl517\bin;C:\Program Files\Microsoft Visual Studio .NET
2003\Common7\IDE;C:\Program Files\Microsoft Visual Studio .NET
2003\VC7\BIN;C:\Program Files\Microsoft Visual Studio .NET
2003\Common7\Tools;C:\Program Files\Microsoft Visual Studio .NET
2003\Common7\Tools\bin\prerelease;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\system32\wbem;
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

See attached patch\. Also attached is the list of formerly exported
things\. Some of them are PerlIO vtables\. I did not test this with GCC
but I don't think anything will fail\. I don't know if Win32 GCC Perl's
miniperl has the same problem as the VC miniperl\. I tested it with plain
C miniperl and a C\+\+ miniperl\, both Visual C 2003 32 bit\. The point of
this patch is to speed up by a tiny bit the Win32 VC Perl build process\.

**Comment by p5pRT** [comment]
### From @bulk88

C&#8203;:\\p517\\perl>dumpbin /exports miniperl\.exe
Microsoft \(R\) COFF/PE Dumper Version 7\.10\.6030
Copyright \(C\) Microsoft Corporation\.  All rights reserved\.


Dump of file miniperl\.exe

File Type&#8203;: EXECUTABLE IMAGE

&nbsp;  Section contains the following exports for miniperl\.exe

&nbsp;    00000000 characteristics
&nbsp;    5075EA6D time date stamp Wed Oct 10 17&#8203;:36&#8203;:45 2012
&nbsp;        0\.00 version
&nbsp;           1 ordinal base
&nbsp;         133 number of functions
&nbsp;         133 number of names

&nbsp;    ordinal hint RVA      name

&nbsp;          1    0 000CB9F0 PerlIO\_byte
&nbsp;          2    1 000CBC90 PerlIO\_crlf
&nbsp;          3    2 000CBC20 PerlIO\_pending
&nbsp;          4    3 000CBBB0 PerlIO\_perlio
&nbsp;          5    4 000CBA60 PerlIO\_raw
&nbsp;          6    5 000CBB40 PerlIO\_stdio
&nbsp;          7    6 000CBAD0 PerlIO\_unix
&nbsp;          8    7 000CB980 PerlIO\_utf8
&nbsp;          9    8 000CCDE0 PerlIO\_win32
&nbsp;         10    9 000AFBEF Perl\_init\_os\_extras
&nbsp;         11    A 000AFE25 Perl\_win32\_init
&nbsp;         12    B 000AFE54 Perl\_win32\_term
&nbsp;         13    C 000AEBF9 win32\_abort
&nbsp;         14    D 000AF223 win32\_access
&nbsp;         15    E 000AE6B6 win32\_alarm
&nbsp;         16    F 000ADA26 win32\_ansipath
&nbsp;         17   10 000AE194 win32\_async\_check
&nbsp;         18   11 000AF9ED win32\_calloc
&nbsp;         19   12 000AF201 win32\_chdir
&nbsp;         20   13 000AF22E win32\_chmod
&nbsp;         21   14 000AF001 win32\_chsize
&nbsp;         22   15 000AF605 win32\_clearenv
&nbsp;         23   16 000AEA9A win32\_clearerr
&nbsp;         24   17 000AF184 win32\_close
&nbsp;         25   18 000ACD7C win32\_closedir
&nbsp;         26   19 000AFFFF win32\_create\_message\_window
&nbsp;         27   1A 000AE735 win32\_crypt
&nbsp;         28   1B 000ACDAC win32\_dirp\_dup
&nbsp;         29   1C 000AF1CD win32\_dup
&nbsp;         30   1D 000AF1D3 win32\_dup2
&nbsp;         31   1E 000AFA83 win32\_dynaload
&nbsp;         32   1F 000AE83D win32\_environ
&nbsp;         33   20 000AF190 win32\_eof
&nbsp;         34   21 000AE837 win32\_errno
&nbsp;         35   22 000AF93C win32\_execv
&nbsp;         36   23 000AF964 win32\_execvp
&nbsp;         37   24 000AEA70 win32\_fclose
&nbsp;         38   25 000AF9B7 win32\_fcloseall
&nbsp;         39   26 000AEA0F win32\_fdopen
&nbsp;         40   27 000AFA0B win32\_fdupopen
&nbsp;         41   28 000AE866 win32\_feof
&nbsp;         42   29 000AE85B win32\_ferror
&nbsp;         43   2A 000AEAA0 win32\_fflush
&nbsp;         44   2B 000AF9C9 win32\_fgetc
&nbsp;         45   2C 000AEB4C win32\_fgetpos
&nbsp;         46   2D 000AF9BD win32\_fgets
&nbsp;         47   2E 000AEA94 win32\_fileno
&nbsp;         48   2F 000AE750 win32\_flock
&nbsp;         49   30 000AF9B1 win32\_flushall
&nbsp;         50   31 000AE9B7 win32\_fopen
&nbsp;         51   32 000AE976 win32\_fprintf
&nbsp;         52   33 000AEA82 win32\_fputc
&nbsp;         53   34 000AEA7C win32\_fputs
&nbsp;         54   35 000AE9AB win32\_fread
&nbsp;         55   36 000AF9F9 win32\_free
&nbsp;         56   37 000AF6B6 win32\_free\_childdir
&nbsp;         57   38 000AF604 win32\_free\_childenv
&nbsp;         58   39 000AF9F9 win32\_freeenvironmentstrings
&nbsp;         59   3A 000AEA46 win32\_freopen
&nbsp;         60   3B 000AEACD win32\_fseek
&nbsp;         61   3C 000AEB52 win32\_fsetpos
&nbsp;         62   3D 000AEBFF win32\_fstat
&nbsp;         63   3E 000AEAA6 win32\_ftell
&nbsp;         64   3F 000AE9B1 win32\_fwrite
&nbsp;         65   40 000AFE7C win32\_get\_child\_IO
&nbsp;         66   41 000AF66A win32\_get\_childdir
&nbsp;         67   42 0009DBB9 win32\_get\_childenv
&nbsp;         68   43 000AFA05 win32\_get\_osfhandle
&nbsp;         69   44 000AEA8E win32\_getc
&nbsp;         70   45 000AF9DB win32\_getchar
&nbsp;         71   46 000ADB8B win32\_getenv
&nbsp;         72   47 000ADB0F win32\_getenvironmentstrings
&nbsp;         73   48 000AC383 win32\_getpid
&nbsp;         74   49 000AF9C3 win32\_gets
&nbsp;         75   4A 000ADF70 win32\_gettimeofday
&nbsp;         76   4B 000B0C92 win32\_ioctl
&nbsp;         77   4C 000AF196 win32\_isatty
&nbsp;         78   4D 000AD345 win32\_kill
&nbsp;         79   4E 000AEE63 win32\_link
&nbsp;         80   4F 000AD7BC win32\_longpath
&nbsp;         81   50 000AF12D win32\_lseek
&nbsp;         82   51 000AF9E7 win32\_malloc
&nbsp;         83   52 000AF1E5 win32\_mkdir
&nbsp;         84   53 000AE24A win32\_msgwait
&nbsp;         85   54 000AF14D win32\_open
&nbsp;         86   55 000AF9FF win32\_open\_osfhandle
&nbsp;         87   56 000AC9D8 win32\_opendir
&nbsp;         88   57 000AC37D win32\_os\_id
&nbsp;         89   58 000AEDE5 win32\_pclose
&nbsp;         90   59 000AF99F win32\_perror
&nbsp;         91   5A 000AEC05 win32\_pipe
&nbsp;         92   5B 000AEC1C win32\_popen
&nbsp;         93   5C 000AEC0B win32\_popenlist
&nbsp;         94   5D 000AE98D win32\_printf
&nbsp;         95   5E 000AF9CF win32\_putc
&nbsp;         96   5F 000AF9E1 win32\_putchar
&nbsp;         97   60 000ADCD9 win32\_putenv
&nbsp;         98   61 000AF9D5 win32\_puts
&nbsp;         99   62 000AF1D9 win32\_read
&nbsp;        100   63 000ACBD9 win32\_readdir
&nbsp;        101   64 000AF9F3 win32\_realloc
&nbsp;        102   65 000AEF5A win32\_rename
&nbsp;        103   66 000AEB58 win32\_rewind
&nbsp;        104   67 000ACD72 win32\_rewinddir
&nbsp;        105   68 000AF1F6 win32\_rmdir
&nbsp;        106   69 000ACD59 win32\_seekdir
&nbsp;        107   6A 000AF9A5 win32\_setbuf
&nbsp;        108   6B 000AEFFB win32\_setmode
&nbsp;        109   6C 000AF9AB win32\_setvbuf
&nbsp;        110   6D 000AFE9F win32\_signal
&nbsp;        111   6E 000AFC2A win32\_signal\_context
&nbsp;        112   6F 000AE66F win32\_sleep
&nbsp;        113   70 000AF6C0 win32\_spawnvp
&nbsp;        114   71 000AD498 win32\_stat
&nbsp;        115   72 000AE843 win32\_stderr
&nbsp;        116   73 000AE84C win32\_stdin
&nbsp;        117   74 000AE852 win32\_stdout
&nbsp;        118   75 000AE8D5 win32\_str\_os\_error
&nbsp;        119   76 000AE871 win32\_strerror
&nbsp;        120   77 000AF147 win32\_tell
&nbsp;        121   78 000ACD47 win32\_telldir
&nbsp;        122   79 000ADD5A win32\_times
&nbsp;        123   7A 000AEB5E win32\_tmpfd
&nbsp;        124   7B 000AEBDF win32\_tmpfile
&nbsp;        125   7C 000ADFC9 win32\_uname
&nbsp;        126   7D 000AEA88 win32\_ungetc
&nbsp;        127   7E 000ADE3C win32\_unlink
&nbsp;        128   7F 000ADEA0 win32\_utime
&nbsp;        129   80 000AE99F win32\_vfprintf
&nbsp;        130   81 000AE9A5 win32\_vprintf
&nbsp;        131   82 000AE661 win32\_wait
&nbsp;        132   83 000AE494 win32\_waitpid
&nbsp;        133   84 000AF1DF win32\_write

&nbsp;  Summary

&nbsp;        4000 \.data
&nbsp;       18000 \.rdata
&nbsp;       B1000 \.text

**Comment by p5pRT** [comment]
### From @bulk88

<details><summary>0001-perl-115216-stop-Win32-VC-miniperl-from-exporting-fu.patch</summary>

```diff
From eec2d6c5c1c0288c9aac407f0e77edf467887f73 Mon Sep 17 00:00:00 2001
From: Daniel Dragan <bulk88@hotmail.com>
Date: Thu, 11 Oct 2012 01:47:19 -0400
Subject: [PATCH] [perl #115216] stop Win32 VC miniperl from exporting
 functions

miniperl.exe does not load XS modules. It has no reason to export anything.
About 130 things are exported by VC Win32 miniperl. 90% of them are
the win32_* functions. All but a couple Perl_* exports are gone in the
exporting miniperl. See perl #115216 for the full list of accidentally
exported items. Also stop trying to find Win32CORE's boot function in
Perl_init_os_extras through the export table. It is not exported and not
in the miniperl image and GetProcAddress will never return not NULL. By
removing this GetProcAddress call, miniperl stops importing GetProcAddress
from kernel32 and a tiny bit startup time by miniperl during the full
perl build process. By removing the exports the compiler is free to use
more random (not cdecl) calling conventions and/or optimizing away code
than before. Also by removing the export entries, and the GetProcAddress
import, RO strings are removed from the miniperl image. This commit only
affects the VC miniperl. The Mingw miniperl remains unmodified except
for not trying to load Win32CORE through the export table and some of the
.c files being compiled with PERL_IS_MINIPERL when previously they were
not.
---
 EXTERN.h          |   22 +++++++++++++++-------
 win32/Makefile    |    2 +-
 win32/makefile.mk |    2 +-
 win32/win32.c     |    2 ++
 win32/win32.h     |   12 ++++++++----
 5 files changed, 27 insertions(+), 13 deletions(-)

diff --git a/EXTERN.h b/EXTERN.h
index 58ca37a..2fd3731 100644
--- a/EXTERN.h
+++ b/EXTERN.h
@@ -29,16 +29,24 @@
 #  define dEXTCONST globaldef {"$GLOBAL_RO_VARS"} readonly
 #else
 #  if (defined(WIN32) || defined(__SYMBIAN32__)) && !defined(PERL_STATIC_SYMS)
-#    if defined(PERLDLL) || defined(__SYMBIAN32__)
-#      define EXT extern __declspec(dllexport)
+/* miniperl should not export anything */
+#    if defined(PERL_IS_MINIPERL) && !defined(UNDER_CE) && defined(_MSC_VER)
+#      define EXT extern
 #      define dEXT 
-#      define EXTCONST extern __declspec(dllexport) const
+#      define EXTCONST extern const
 #      define dEXTCONST const
 #    else
-#      define EXT extern __declspec(dllimport)
-#      define dEXT 
-#      define EXTCONST extern __declspec(dllimport) const
-#      define dEXTCONST const
+#      if defined(PERLDLL) || defined(__SYMBIAN32__)
+#        define EXT extern __declspec(dllexport)
+#        define dEXT 
+#        define EXTCONST extern __declspec(dllexport) const
+#        define dEXTCONST const
+#      else
+#        define EXT extern __declspec(dllimport)
+#        define dEXT 
+#        define EXTCONST extern __declspec(dllimport) const
+#       define dEXTCONST const
+#     endif
 #    endif
 #  else
 #    if defined(__CYGWIN__) && defined(USEIMPORTLIB)
diff --git a/win32/Makefile b/win32/Makefile
index f005286..cf83e9b 100644
--- a/win32/Makefile
+++ b/win32/Makefile
@@ -968,7 +968,7 @@ $(MINICORE_OBJ) : $(CORE_NOCFG_H)
 	$(CC) -c $(CFLAGS) -DPERL_EXTERNAL_GLOB -DPERL_IS_MINIPERL $(OBJOUT_FLAG)$@ ..\$(*F).c
 
 $(MINIWIN32_OBJ) : $(CORE_NOCFG_H)
-	$(CC) -c $(CFLAGS) $(OBJOUT_FLAG)$@ $(*F).c
+	$(CC) -c $(CFLAGS) -DPERL_IS_MINIPERL $(OBJOUT_FLAG)$@ $(*F).c
 
 # -DPERL_IMPLICIT_SYS needs C++ for perllib.c
 # This is the only file that depends on perlhost.h, vmem.h, and vdir.h
diff --git a/win32/makefile.mk b/win32/makefile.mk
index e9d40f8..bd8e726 100644
--- a/win32/makefile.mk
+++ b/win32/makefile.mk
@@ -1097,7 +1097,7 @@ $(MINICORE_OBJ) : $(CORE_NOCFG_H)
 	$(CC) -c $(CFLAGS) -DPERL_EXTERNAL_GLOB -DPERL_IS_MINIPERL $(OBJOUT_FLAG)$@ ..\$(*B).c
 
 $(MINIWIN32_OBJ) : $(CORE_NOCFG_H)
-	$(CC) -c $(CFLAGS) $(OBJOUT_FLAG)$@ $(*B).c
+	$(CC) -c $(CFLAGS) -DPERL_IS_MINIPERL $(OBJOUT_FLAG)$@ $(*B).c
 
 # -DPERL_IMPLICIT_SYS needs C++ for perllib.c
 # rules wrapped in .IFs break Win9X build (we end up with unbalanced []s unless
diff --git a/win32/win32.c b/win32/win32.c
index 4427e06..3ee85a1 100644
--- a/win32/win32.c
+++ b/win32/win32.c
@@ -4146,10 +4146,12 @@ Perl_init_os_extras(void)
     char *file = __FILE__;
 
     /* Initialize Win32CORE if it has been statically linked. */
+#ifndef PERL_IS_MINIPERL
     void (*pfn_init)(pTHX);
     pfn_init = (void (*)(pTHX))GetProcAddress((HMODULE)w32_perldll_handle, "init_Win32CORE");
     if (pfn_init)
         pfn_init(aTHX);
+#endif
 
     newXS("Win32::SetChildShowWindow", w32_SetChildShowWindow, file);
 }
diff --git a/win32/win32.h b/win32/win32.h
index 59a3052..9d229db 100644
--- a/win32/win32.h
+++ b/win32/win32.h
@@ -46,11 +46,15 @@
  */
 
 /* now even GCC supports __declspec() */
-
-#if defined(PERLDLL)
-#define DllExport __declspec(dllexport)
+/* miniperl has no reason to export anything */
+#if defined(PERL_IS_MINIPERL) && !defined(UNDER_CE) && defined(_MSC_VER)
+#  define DllExport
 #else
-#define DllExport __declspec(dllimport)
+#  if defined(PERLDLL)
+#    define DllExport __declspec(dllexport)
+#  else
+#    define DllExport __declspec(dllimport)
+#  endif
 #endif
 
 /* The Perl APIs can only be called directly inside the perl5xx.dll.
-- 
1.7.9.msysgit.0


```
</details>

**Comment by p5pRT** [comment]
@bulk88 - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @bulk88

By removing the exports\, some of the win32\_\* functions are removed from
the image\, for example win32\_dynaload is never called from within
miniperl and it disappears from the image if its export is removed\.

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Wed Oct 10 23&#8203;:49&#8203;:16 2012\, bulk88 wrote&#8203;:
> By removing the exports\, some of the win32\_\* functions are removed from
> the image\, for example win32\_dynaload is never called from within
> miniperl and it disappears from the image if its export is removed\.

Thanks\, applied as a19baa613c \(with indentation fixes in def6ed225c\)\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
